contracts/Marketplace1155.sol

```diff
@@ -0,0 +1,161 @@
+    ) public virtual override returns (bytes4) {
+        return this.onERC1155BatchReceived.selector;
+    }
+
+    function createMarketItem(
+        address nftContract,
+        uint256[] memory tokenIds,
+        uint256 price,
+        uint256[] memory amounts // se e' un nft amounts non dovrebbe essere sempre 1? siccome immagino siano tracked dai tokenIds questo mi sembra superfluo
+    ) public payable nonReentrant {
+        require(price > 0, "Price must be greater than 0"); // Non vogliamo controllare che tokenIds sia lungo quanto amounts?
+
+        IERC1155(nftContract).safeBatchTransferFrom(
+            msg.sender,
+            address(this),
+            tokenIds,
+            amounts,
+            ""
+        );
+
+        for (uint256 i = 0; i <= (tokenIds.length - 1); i++) {
+            //davide: possiamo usare solo il < e togliere il -1?
+            _itemIds.increment();
+            uint256 itemId = _itemIds.current();
+
+            idToMarketItem[itemId] = MarketItem(
+                itemId,
+                nftContract,
+                tokenIds[i],
+                payable(msg.sender),
+                payable(address(0)),
+                price,
+                false
+            );
+
+            emit MarketItemCreated(
+                itemId,
+                nftContract,
+                tokenIds[i],
+                msg.sender,
+                address(0),
+                price,
+                false
+            );
+
+            // non vale la pena emettere un evento solo invece di n?
+        }
+    }
+
+    function createMarketSale(
+        address nftContract,
+        uint256 itemId,
+        uint256[] memory _tokenIds,
+        uint256[] memory amounts
+    ) public payable nonReentrant {
+        uint256 price = idToMarketItem[itemId].price;
+
+        bool sold = idToMarketItem[itemId].sold;
+        require(
+            msg.value == price,
+            "Please submit the asking price in order to complete the purchase"
+        );
+        require(sold != true, "This Sale has alredy finished"); //missing a
+        emit MarketItemSold(itemId, msg.sender); //Possiamo metterlo alla fine questo?
+
+        idToMarketItem[itemId].seller.transfer(msg.value);
+        IERC1155(nftContract).safeBatchTransferFrom(
+            address(this),
+            msg.sender,
+            _tokenIds,
+            amounts,
+            ""
+        );
+        idToMarketItem[itemId].owner = payable(msg.sender);
+        _itemsSold.increment();
+        idToMarketItem[itemId].sold = true;
+    }
+
+    function fetchMarketItems() public view returns (MarketItem[] memory) {
+        uint256 itemCount = _itemIds.current();
+        uint256 unsoldItemCount = _itemIds.current() - _itemsSold.current();
+        uint256 currentIndex = 0;
+
+        MarketItem[] memory items = new MarketItem[](unsoldItemCount);
+        for (uint256 i = 0; i < itemCount; i++) {
+            if (idToMarketItem[i + 1].owner == address(0)) {
```

Comment:
Why the element `0` is skipped?
Despite of this, the currentId can be easily removed.
If we start from element `1` the code can be rewritten like this:

```
...
        for (uint256 i = 1; i < itemCount; i++) {
            if (idToMarketItem[i].owner == address(0)) {
                MarketItem storage currentItem = idToMarketItem[i];
                items[i] = currentItem;
            }
        }
...
```

Reply:
`idToMarketItem[i + 1].owner == address(0)` we should change this in case we decide to use the bool `sold`